packages/actor-init-sparql/lib/HttpServiceSparqlEndpoint.ts

```diff
@@ -158,8 +159,48 @@ Options:
   public async handleRequest(engine: ActorInitSparql, variants: { type: string, quality: number }[],
                              stdout: Writable, stderr: Writable,
                              request: http.IncomingMessage, response: http.ServerResponse) {
-    const mediaType: string = request.headers.accept && request.headers.accept !== '*/*'
-      ? require('negotiate').choose(variants, request)[0].type : null;
+
+    // Negotiate the best mediatype format
+    let negotiation = undefined;
+    let mediaType = null;
+    try {
+
+      negotiation = require('negotiate').choose(variants, request);
+      const isValid = request.headers.accept && request.headers.accept !== '*/*' && negotiation;
+      mediaType = isValid ? negotiation[0].type : null;
+
+      if (isValid && negotiation.length > 1) {
+        // The preferred negotiation
+        let finalNegotiation = negotiation[0];
+        // Best quality achievable
+        const quality = negotiation[0].q;
+        let found = false;
+        let k = 0;
+        // Iterate through all the variants in order
+        while (!found) {
```

Comment:
I'm wondering if this algorithm can be made more efficient, as it's O(n^2) at the moment.
Wouldn't you just need to find the negotiation with the highest q?
If so, this could even be a one-liner.

Reply:
We have the negotiation with the highest quality, but it isn't necessarily unique, so finding the highest quality doesn't give the default preferred writer. we would for example have gotten a list from the negotiation as follows: [{type:stats: quality:1},{type:application/json: quality:1},{type:simple: quality:1}] However we prefer it as follows (given with the variants: [{type:application/json: quality:1},{type:simple: quality:1},{type:stats: quality:1}]. Then we would need to check if the application/json has the same quality given by the negotiator as the given negotiation[0]. And if not check for the second preferred variant. The main issue is that the negotiate library doesn't keep the ordering of same quality elements

Reply:
Ah ok, got it.

Instead of having a double loop for this, wouldn't it be more efficient to just iterate over _all_ original variants (ordered by preference), where the first one that is found in the negotiated list is returned?

Reply:
the negotiated list gives back all qualities. So if it were to be 1 as best it would have a list [1,1,1,0.9,0.5,0.4]. Meaning that if contains wouldn't work. Wouldn't your suggestion also be a double loop internally? Contains is O(n) then n time in worst case would also be O(n^2)?

Reply:
I don't see why containment checking wouldn't work. The returned list _also_ contains the mediatype labels, right? So I would just compare by them.

And you're right about complexity, that would be the same. However, readability would be improved, since a lot less code would needed.